lib/symbConstant.ml

```diff
@@ -77,11 +87,17 @@ module Make(Scalar:Scalar.S) = struct
       Misc.string_eq  s1 s2 && Misc.int_eq p1 p2
   | Tag t1,Tag t2 -> Misc.string_eq t1 t2
   | PteVal p1,PteVal p2 -> PTEVal.compare p1 p2 = 0
-  | (PteVal _,(Symbolic _|Concrete _|Label _|Tag _))
-  | (Concrete _,(Symbolic _|Label _|Tag _|PteVal _))
-  | (Symbolic _,(Concrete _|Label _|Tag _|PteVal _))
-  | (Label _,(Concrete _|Symbolic _|Tag _|PteVal _))
-  | (Tag _,(Concrete _|Symbolic _|Label _|PteVal _))
+  | MemAttr l1,MemAttr l2 ->
+      List.length l1 = List.length l2 &&
+        List.fold_right2
+          (fun s1 s2 c -> c && Misc.string_eq s1 s2)
+          l1 l2 true
```

Comment:
Is there some assumed sorting of attributes?
If not `["A";"B"]` and `["B";"A";]` are different, if yes they are identical and the above comparison function is wrong.

In any case an appropriate comment near the definition of `MemAttr` would be useful.

Reply:
Moreover, if sorted, the list of strings should perhaps be implemented as an abstract data type, so as to easily maintain the invariant.

Reply:
That's a good point and I will implement it as a sorted list of strings. At some point, I think we need to add sanity checks, those sanity checks would be different for different architectures. For example, for  AArch64 a location can be either in Normal or in Device Memory, as well as defaults if we have specified a memory type, but we have other attributes e.g., cacheablility, then it should be Normal memory. Not sure if those checks fit better in the cat file or in herd.

Reply:
hi Nikos, personally I would put those sanity checks in the cat file.
Thanks,
Jade

Reply:
Let us follow @jalglave  here.